AESinkAudioTrack: Simplify RAW - buffer did not work anyways

fritsch opened the pull request:
## Description
I looked a bit at Nova-Player. Seems the colleagues over there simplified the RAW handling quite a lot. I adjusted accordingly.

Let's just use 32K buffer for NON-HD formats, 128 for HR formats as well as 256K for HD formats and four periods to not cause underruns immediately. This makes as much sense as the other values.
    
Tested with all formats available on my FireTV 4K Max 2nd Gen. This device cannot do TrueHD via Android's Packer.

## Motivation and context
When Nova-Player works, we should also work - and follow their guess-work.

## How has this been tested?
I tested the formats my device supports in RAW mode: AC3, EAC3, DTS-HD-MA, DTS - it does not support TrueHD via the RAW API.

## What is the effect on users?
Some people might have better luck with passthrough and especially seeking when they have to use the RAW passthrough method.

fritsch commented on xbmc/cores/AudioEngine/Sinks/AESinkAUDIOTRACK.cpp:
```diff
@@ -415,57 +413,31 @@ bool CAESinkAUDIOTRACK::Initialize(AEAudioFormat &format, std::string &device)
-          m_format.m_frames = m_min_buffer_size; // needs testing
-          rawlength_in_seconds = 8 * m_format.m_streamInfo.GetDuration() / 1000;
+          m_min_buffer_size = std::max(m_min_buffer_size * 2, MAX_RAW_AUDIO_BUFFER);
+          m_format.m_frames = m_min_buffer_size / 2;
```
We don't. Please see above.

thexai commented:
Tested on Shield (RAW only) and i get mixed results:

TrueHD has improved a lot in seeks.
DTS-HD MA is worse in seeks.

EAC3, AC3: is hard to say in general works fine (also before) but encountered one rare case that fails even normal playback:

EAC3 two channels

```
Audio #1
ID                                       : 2
Format                                   : E-AC-3
Format/Info                              : Enhanced AC-3
Commercial name                          : Dolby Digital Plus
Codec ID                                 : A_EAC3
Duration                                 : 1 h 59 min
Bit rate mode                            : Constant
Bit rate                                 : 224 kb/s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 48.0 kHz
Frame rate                               : 31.250 FPS (1536 SPF)
Compression mode                         : Lossy
Stream size                              : 191 MiB (3%)
Language                                 : Spanish
Service kind                             : Complete Main
Default                                  : Yes
Forced                                   : Yes
Dialog Normalization                     : -31 dB
compr                                    : -0.28 dB
dsurmod                                  : Not Dolby Surround encoded
dialnorm_Average                         : -31 dB
dialnorm_Minimum                         : -31 dB
dialnorm_Maximum                         : -31 dB
```

https://paste.kodi.tv/zaxiduxedi.kodi

I think that as it is, more regressions may occur. Perhaps it could be used to improve TrueHD RAW only. 

Or it will require much more testing in forums.

fritsch commented:
@thexai could you retest? I force-pushed in the middle as I had time to finetune a bit more on my system. Seems the EAC3 you tested still used the 32K buffer while it should use the 64K version 16K periods?

See: https://github.com/xbmc/xbmc/pull/27621/commits/65c49252db9ac559c4ff636b0bf3133a55450d34#diff-9884fdeb14fb2a0179b4983db00f524e1298005123d633a3b2ed9a261248e1e9R438 

This also changes the ratio. Can you retry EAC3 and DTS-HD-MA?

For the 2 channel EAC3 - did it work before?

fritsch committed:
AESinkAUDIOTRACK: Improve TrueHD and DTS-HD configuration

Use two periods for TrueHD.
Adjust DTS-HD to an empirical value of 30720 * 4 for the buffers

fritsch commented:
Most stuff does not really make sense ... especially the very moment the dynamic subframes come into play ... ;-) but we knew that before.

thexai commented:
Tested https://github.com/fritsch/xbmc/commits/at2025final2/

TrueHD works well also with buffer of two frames. Pause commit seems helps also and now seeks are almost perfect 95% of times.

fritsch commented:
What about DTS-HD and such? Sadly the code is still mathematically not yet correct - the thing with AddPause is, that it is used for multiple disciplines ... warm up, sleep out, etc. and while the sink is paused ... we cannot really sleep out properly ...

It needs to act like this:
Sink is empty: Pause bursts fill up until limit is reached - sink blocks. Real data starts.
Sink is already filled, but AE wants to move buffer a bit ahead / e.g. out. That way it stiffs pause bursts in to move audio ahead.

Especially the second one is tricky ... to get it right.

thexai commented on xbmc/cores/AudioEngine/Sinks/AESinkAUDIOTRACK.cpp:
```diff
@@ -422,14 +421,14 @@ bool CAESinkAUDIOTRACK::Initialize(AEAudioFormat &format, std::string &device)
-          rawlength_in_seconds = 8 * m_format.m_streamInfo.GetDuration() / 1000; // on average
+          m_min_buffer_size = MAX_RAW_AUDIO_BUFFER_TRUEHD;
+          m_format.m_frames = m_min_buffer_size / 2;
+          rawlength_in_seconds = 0.04;
```
To avoid magic numbers seems better replace 0.04 with:

`2 * m_format.m_streamInfo.GetDuration() / 1000;`

Also for me seems more clear simply define only 61440 number. I mean:

```c++
const unsigned int TRUEHD_SIZE = 61440;

m_min_buffer_size = 2 * TRUEHD_SIZE;
m_format.m_frames = TRUEHD_SIZE;
rawlength_in_seconds = 2 * m_format.m_streamInfo.GetDuration() / 1000;
```

With this is clear that is used a buffer two times TrueHD size.

fritsch replied:
I continue on at2025final2 - will ping you again when I am happy - then we can take over the bits we want in here. Thanks again.

thexai replied:
Okay, I'll continue tomorrow.

fritsch committed:
AESinkAudioTrack: Revisit pause burst

In the great endeveour happening every time around christmas I try to find
sweet spots for my device here at home to finally make DTS-HD seeking working.

I fail every time ;-). But this time it improved situation for Nvidia Shield
a bit further in RAW mode. DTS-HD-MA seeking now seems to work nicely.

fritsch committed:
Stop using Silence for DTS-HD and DTS-HD_MA

Last resort ...

fritsch committed:
AESinkAudioTrack: Stop unneeded resize

The buffers are resized whenever the input bytes change. This is especially
causing many reallocations / ptr changes when using RAW PT where the byte
size of the frames can vary per AddPacket. Stop doing this.

thexai reviewed (approved):
Tested with the latest update. The commit "no silence on pause for DTS-HD / MA" supposes a big improvement from user's perspective (it may be a hack but works) not only improves the pause but also seek is 100% perfect now.

thexai commented:
Yes, it's better open another PR for the getLatency changes.
